File: src/ultra_lean_mcp_proxy/result_compression.py
"""Generic structured JSON result compression for Ultra Lean MCP Proxy."""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from typing import Any, MutableMapping


@dataclass
class CompressionOptions:
    mode: str = "balanced"  # off | balanced | aggressive
    strip_nulls: bool = False
    strip_defaults: bool = False
    min_payload_bytes: int = 512
    enable_columnar: bool = True
    columnar_min_rows: int = 8
    columnar_min_fields: int = 2


def _json_size(value: Any) -> int:
    return len(json.dumps(value, separators=(",", ":"), ensure_ascii=False))


def _stable_json(value: Any) -> str:
    return json.dumps(value, separators=(",", ":"), ensure_ascii=False, sort_keys=True)


class TokenCounter:
    """Best-effort token estimator with optional tiktoken backend."""

    def __init__(self, encoding_name: str = "cl100k_base", *, strict: bool = False):
        self._enc = None
        self.encoding_name = encoding_name
        self.backend = "heuristic"
        self.reason = "tiktoken_unavailable_or_encoding_missing"
        try:
            import tiktoken  # type: ignore
        except ImportError as exc:
            if strict:
                raise ValueError(
                    "tiktoken is required for strict token counting but is not installed"
                ) from exc
            tiktoken = None  # type: ignore
        if tiktoken:
            try:
                self._enc = tiktoken.get_encoding(encoding_name)
                self.backend = "tiktoken"
                self.reason = "ok"
            except Exception as exc:
                if strict:
                    raise ValueError(
                        f"Requested tokenizer encoding '{encoding_name}' is unavailable"
                    ) from exc
                self._enc = None

    def count(self, value: Any) -> int:
        text = json.dumps(value, separators=(",", ":"), ensure_ascii=False)
        if self._enc is not None:
            return len(self._enc.encode(text))
        # Deterministic fallback approximation.
        return max(1, len(text) // 4)


def _collect_key_frequency(node: Any, counter: dict[str, int]):
    if isinstance(node, dict):
        for key, value in node.items():
            counter[str(key)] = counter.get(str(key), 0) + 1
            _collect_key_frequency(value, counter)
    elif isinstance(node, list):
        for item in node:
            _collect_key_frequency(item, counter)


def _build_key_aliases(counter: dict[str, int], mode: str) -> dict[str, str]:
    if mode == "off":
        return {}
    min_freq = 1 if mode == "aggressive" else 2
    candidates = [
        (key, freq)
        for key, freq in counter.items()
        if freq >= min_freq and len(key) > 2
    ]
    # Rank by frequency then length: prioritize repetitive long keys.
    candidates.sort(key=lambda x: (x[1], len(x[0])), reverse=True)

    aliases: dict[str, str] = {}
    for idx, (key, _) in enumerate(candidates):
        alias = f"k{idx}"
        if len(alias) < len(key):
            aliases[key] = alias
    return aliases


def _is_defaultish(value: Any) -> bool:
    return value in (None, "", 0, False, [], {})


def _can_columnar(items: list[Any], opts: CompressionOptions) -> tuple[bool, list[str]]:
    if not opts.enable_columnar:
        return False, []
    if len(items) < opts.columnar_min_rows:
        return False, []
    if not items or not all(isinstance(item, dict) for item in items):
        return False, []

    first_keys = list(items[0].keys())
    if len(first_keys) < opts.columnar_min_fields:
        return False, []

    first_set = set(first_keys)
    for item in items[1:]:
        if set(item.keys()) != first_set:
            return False, []
    return True, first_keys


def _encode(node: Any, key_alias: dict[str, str], opts: CompressionOptions) -> Any:
    if isinstance(node, dict):
        out = {}
        for key, value in node.items():
            if opts.strip_nulls and value is None:
                continue
            if opts.strip_defaults and str(key).lower() in {"default", "defaults"} and _is_defaultish(value):
                continue
            encoded_key = key_alias.get(str(key), str(key))
            out[encoded_key] = _encode(value, key_alias, opts)
        return out

    if isinstance(node, list):
        can_col, columns = _can_columnar(node, opts)
        if can_col:
            encoded_columns = [key_alias.get(str(col), str(col)) for col in columns]
            encoded_rows = []
            for item in node:
                row = [_encode(item[col], key_alias, opts) for col in columns]
                encoded_rows.append(row)
            return {"~t": {"c": encoded_columns, "r": encoded_rows}}
        return [_encode(item, key_alias, opts) for item in node]

    return node


def _decode(node: Any, alias_to_key: dict[str, str]) -> Any:
    if isinstance(node, dict):
        if "~t" in node and isinstance(node["~t"], dict):
            meta = node["~t"]
            columns = meta.get("c", [])
            rows = meta.get("r", [])
            if isinstance(columns, list) and isinstance(rows, list):
                decoded_items = []
                decoded_cols = [alias_to_key.get(str(col), str(col)) for col in columns]
                for row in rows:
                    if not isinstance(row, list):
                        continue
                    obj = {}
                    for idx, col in enumerate(decoded_cols):
                        if idx < len(row):
                            obj[col] = _decode(row[idx], alias_to_key)
                    decoded_items.append(obj)
                return decoded_items

        out = {}
        for key, value in node.items():
            decoded_key = alias_to_key.get(str(key), str(key))
            out[decoded_key] = _decode(value, alias_to_key)
        return out

    if isinstance(node, list):
        return [_decode(item, alias_to_key) for item in node]

    return node


def _key_ref(alias_to_key: dict[str, str]) -> str:
    digest = hashlib.sha256(_stable_json(alias_to_key).encode("utf-8")).hexdigest()[:12]
    return f"kdict-{digest}"


def compress_result(
    input_data: Any,
    options: CompressionOptions | None = None,
    *,
    key_registry: MutableMapping[str, dict[str, str]] | None = None,
    registry_counter: MutableMapping[str, int] | None = None,
    reuse_keys: bool = False,
    key_bootstrap_interval: int = 8,
) -> dict[str, Any]:
    """Compress structured JSON result with generic reversible transforms."""
    opts = options or CompressionOptions()
    original_bytes = _json_size(input_data)
    if original_bytes < opts.min_payload_bytes:
        return {
            "encoding": "lapc-json-v1",
            "compressed": False,
            "originalBytes": original_bytes,
            "compressedBytes": original_bytes,
            "savedBytes": 0,
            "savedRatio": 0.0,
            "data": input_data,
            "keys": {},
        }

    key_counter: dict[str, int] = {}
    _collect_key_frequency(input_data, key_counter)
    key_alias = _build_key_aliases(key_counter, opts.mode)
    encoded = _encode(input_data, key_alias, opts)
    alias_to_key = {alias: key for key, alias in key_alias.items()}

    envelope: dict[str, Any] = {
        "encoding": "lapc-json-v1",
        "compressed": True,
        "mode": opts.mode,
        "originalBytes": original_bytes,
        "data": encoded,
        "keys": alias_to_key,
    }

    if reuse_keys and key_registry is not None:
        ref = _key_ref(alias_to_key)
        include_keys = True
        previous = key_registry.get(ref)
        if previous == alias_to_key:
            include_keys = False
            if registry_counter is not None:
                count = registry_counter.get(ref, 0) + 1
                registry_counter[ref] = count
                if key_bootstrap_interval > 0 and (count % key_bootstrap_interval) == 0:
                    include_keys = True
        else:
            key_registry[ref] = dict(alias_to_key)
            if registry_counter is not None:
                registry_counter[ref] = 1

        envelope["keysRef"] = ref
        if not include_keys:
            envelope.pop("keys", None)

    compressed_bytes = _json_size(envelope)
    saved = original_bytes - compressed_bytes
    envelope["compressedBytes"] = compressed_bytes
    envelope["savedBytes"] = saved
    envelope["savedRatio"] = (saved / original_bytes) if original_bytes else 0.0

    if saved <= 0:
        envelope["compressed"] = False
        envelope["data"] = input_data
        envelope["keys"] = {}
        envelope.pop("keysRef", None)
        envelope["compressedBytes"] = original_bytes
        envelope["savedBytes"] = 0
        envelope["savedRatio"] = 0.0
    return envelope


def decompress_result(
    envelope: dict[str, Any],
    *,
    key_registry: MutableMapping[str, dict[str, str]] | None = None,
) -> Any:
    """Decompress a result previously produced by ``compress_result``."""
    if not isinstance(envelope, dict) or envelope.get("encoding") != "lapc-json-v1":
        raise ValueError("Unsupported compression envelope")
    data = envelope.get("data")
    if not envelope.get("compressed"):
        return data

    keys = envelope.get("keys")
    if not isinstance(keys, dict):
        keys = None

    keys_ref = envelope.get("keysRef")
    if keys is None and isinstance(keys_ref, str) and key_registry is not None:
        keys = key_registry.get(keys_ref)

    if not isinstance(keys, dict):
        raise ValueError("Invalid or missing key dictionary in envelope")

    return _decode(data, keys)


def token_savings(original: Any, candidate: Any, counter: TokenCounter | None = None) -> int:
    """Return positive value when candidate uses fewer tokens than original."""
    c = counter or TokenCounter()
    return c.count(original) - c.count(candidate)


def estimate_compressibility(value: Any) -> float:
    """Estimate whether a payload is likely to benefit from structural compression.

    Returns a score in `[0, 1]` based on key repetition, duplicate scalar values,
    and homogeneous list-of-object shapes.
    """
    key_counter: dict[str, int] = {}
    scalar_counter: dict[str, int] = {}
    homogeneous_lists = 0
    total_lists = 0

    def walk(node: Any):
        nonlocal homogeneous_lists, total_lists
        if isinstance(node, dict):
            for key, child in node.items():
                k = str(key)
                key_counter[k] = key_counter.get(k, 0) + 1
                walk(child)
        elif isinstance(node, list):
            total_lists += 1
            if node and all(isinstance(item, dict) for item in node):
                keysets = [tuple(sorted(item.keys())) for item in node]
                if len(set(keysets)) == 1:
                    homogeneous_lists += 1
            for child in node:
                walk(child)
        else:
            if isinstance(node, (str, int, float, bool)) or node is None:
                marker = json.dumps(node, ensure_ascii=False)
                scalar_counter[marker] = scalar_counter.get(marker, 0) + 1

    walk(value)

    total_keys = sum(key_counter.values())
    duplicate_keys = max(0, total_keys - len(key_counter))
    key_repeat_ratio = (duplicate_keys / total_keys) if total_keys else 0.0

    total_scalars = sum(scalar_counter.values())
    duplicate_scalars = max(0, total_scalars - len(scalar_counter))
    scalar_repeat_ratio = (duplicate_scalars / total_scalars) if total_scalars else 0.0

    homogeneous_ratio = (homogeneous_lists / total_lists) if total_lists else 0.0

    score = 0.5 * key_repeat_ratio + 0.25 * scalar_repeat_ratio + 0.25 * homogeneous_ratio
    if score < 0:
        return 0.0
    if score > 1:
        return 1.0
    return score

File: npm/src/result-compression.mjs
/**
 * Generic structured JSON result compression for Ultra Lean MCP Proxy.
 */

import { createHash } from 'node:crypto';
import { cloneJson } from './state.mjs';

function jsonSize(value) {
  return Buffer.byteLength(JSON.stringify(value), 'utf-8');
}

function stableJson(value) {
  return JSON.stringify(value, Object.keys(value || {}).sort());
}

export function makeCompressionOptions(overrides = {}) {
  return {
    mode: 'balanced', // off | balanced | aggressive
    stripNulls: false,
    stripDefaults: false,
    minPayloadBytes: 512,
    enableColumnar: true,
    columnarMinRows: 8,
    columnarMinFields: 2,
    ...overrides,
  };
}

export class TokenCounter {
  constructor() {
    this.backend = 'heuristic';
    this.reason = 'node_builtin_estimator';
  }

  count(value) {
    const text = JSON.stringify(value);
    return Math.max(1, Math.floor(text.length / 4));
  }
}

function collectKeyFrequency(node, counter) {
  if (Array.isArray(node)) {
    for (const item of node) {
      collectKeyFrequency(item, counter);
    }
    return;
  }
  if (node && typeof node === 'object') {
    for (const [key, value] of Object.entries(node)) {
      const name = String(key);
      counter[name] = (counter[name] || 0) + 1;
      collectKeyFrequency(value, counter);
    }
  }
}

function buildKeyAliases(counter, mode) {
  if (mode === 'off') return {};
  const minFreq = mode === 'aggressive' ? 1 : 2;
  const candidates = Object.entries(counter)
    .filter(([key, freq]) => freq >= minFreq && key.length > 2)
    .sort((a, b) => {
      if (a[1] !== b[1]) return b[1] - a[1];
      return b[0].length - a[0].length;
    });

  const aliases = {};
  for (let i = 0; i < candidates.length; i++) {
    const key = candidates[i][0];
    const alias = `k${i}`;
    if (alias.length < key.length) {
      aliases[key] = alias;
    }
  }
  return aliases;
}

function isDefaultish(value) {
  if (value === null || value === '' || value === 0 || value === false) return true;
  if (Array.isArray(value) && value.length === 0) return true;
  if (value && typeof value === 'object' && Object.keys(value).length === 0) return true;
  return false;
}

function canColumnar(items, opts) {
  if (!opts.enableColumnar) return [false, []];
  if (!Array.isArray(items) || items.length < opts.columnarMinRows) return [false, []];
  if (!items.every((item) => item && typeof item === 'object' && !Array.isArray(item))) {
    return [false, []];
  }
  const firstKeys = Object.keys(items[0]);
  if (firstKeys.length < opts.columnarMinFields) return [false, []];
  const firstSet = new Set(firstKeys);
  for (let i = 1; i < items.length; i++) {
    const keys = Object.keys(items[i]);
    if (keys.length !== firstSet.size) return [false, []];
    if (keys.some((key) => !firstSet.has(key))) return [false, []];
  }
  return [true, firstKeys];
}

function encode(node, keyAlias, opts) {
  if (Array.isArray(node)) {
    const [asColumnar, columns] = canColumnar(node, opts);
    if (asColumnar) {
      const encodedColumns = columns.map((col) => keyAlias[col] || col);
      const rows = [];
      for (const item of node) {
        rows.push(columns.map((col) => encode(item[col], keyAlias, opts)));
      }
      return { '~t': { c: encodedColumns, r: rows } };
    }
    return node.map((item) => encode(item, keyAlias, opts));
  }
  if (node && typeof node === 'object') {
    const out = {};
    for (const [key, value] of Object.entries(node)) {
      if (opts.stripNulls && value === null) continue;
      if (
        opts.stripDefaults
        && ['default', 'defaults'].includes(String(key).toLowerCase())
        && isDefaultish(value)
      ) {
        continue;
      }
      const encodedKey = keyAlias[String(key)] || String(key);
      out[encodedKey] = encode(value, keyAlias, opts);
    }
    return out;
  }
  return node;
}

function decode(node, aliasToKey) {
  if (Array.isArray(node)) {
    return node.map((item) => decode(item, aliasToKey));
  }
  if (node && typeof node === 'object') {
    if (node['~t'] && typeof node['~t'] === 'object') {
      const meta = node['~t'];
      const columns = Array.isArray(meta.c) ? meta.c : [];
      const rows = Array.isArray(meta.r) ? meta.r : [];
      const decodedColumns = columns.map((col) => aliasToKey[String(col)] || String(col));
      const items = [];
      for (const row of rows) {
        if (!Array.isArray(row)) continue;
        const item = {};
        for (let i = 0; i < decodedColumns.length; i++) {
          if (i < row.length) {
            item[decodedColumns[i]] = decode(row[i], aliasToKey);
          }
        }
        items.push(item);
      }
      return items;
    }
    const out = {};
    for (const [key, value] of Object.entries(node)) {
      const decodedKey = aliasToKey[String(key)] || String(key);
      out[decodedKey] = decode(value, aliasToKey);
    }
    return out;
  }
  return node;
}

function keyRef(aliasToKey) {
  const digest = createHash('sha256').update(stableJson(aliasToKey), 'utf-8').digest('hex').slice(0, 12);
  return `kdict-${digest}`;
}

export function compressResult(
  inputData,
  options = null,
  {
    keyRegistry = null,
    registryCounter = null,
    reuseKeys = false,
    keyBootstrapInterval = 8,
  } = {}
) {
  const opts = makeCompressionOptions(options || {});
  const originalBytes = jsonSize(inputData);

  if (originalBytes < opts.minPayloadBytes) {
    return {
      encoding: 'lapc-json-v1',
      compressed: false,
      originalBytes,
      compressedBytes: originalBytes,
      savedBytes: 0,
      savedRatio: 0,
      data: inputData,
      keys: {},
    };
  }

  const keyCounter = {};
  collectKeyFrequency(inputData, keyCounter);
  const keyAlias = buildKeyAliases(keyCounter, opts.mode);
  const encoded = encode(inputData, keyAlias, opts);
  const aliasToKey = {};
  for (const [key, alias] of Object.entries(keyAlias)) {
    aliasToKey[alias] = key;
  }

  const envelope = {
    encoding: 'lapc-json-v1',
    compressed: true,
    mode: opts.mode,
    originalBytes,
    data: encoded,
    keys: aliasToKey,
  };

  if (reuseKeys && keyRegistry && typeof keyRegistry === 'object') {
    const ref = keyRef(aliasToKey);
    let includeKeys = true;
    const previous = keyRegistry[ref];
    if (previous && JSON.stringify(previous) === JSON.stringify(aliasToKey)) {
      includeKeys = false;
      if (registryCounter && typeof registryCounter === 'object') {
        const count = (registryCounter[ref] || 0) + 1;
        registryCounter[ref] = count;
        if (keyBootstrapInterval > 0 && count % keyBootstrapInterval === 0) {
          includeKeys = true;
        }
      }
    } else {
      keyRegistry[ref] = cloneJson(aliasToKey);
      if (registryCounter && typeof registryCounter === 'object') {
        registryCounter[ref] = 1;
      }
    }
    envelope.keysRef = ref;
    if (!includeKeys) {
      delete envelope.keys;
    }
  }

  const compressedBytes = jsonSize(envelope);
  const savedBytes = originalBytes - compressedBytes;
  envelope.compressedBytes = compressedBytes;
  envelope.savedBytes = savedBytes;
  envelope.savedRatio = originalBytes > 0 ? savedBytes / originalBytes : 0;

  if (savedBytes <= 0) {
    envelope.compressed = false;
    envelope.data = inputData;
    envelope.keys = {};
    delete envelope.keysRef;
    envelope.compressedBytes = originalBytes;
    envelope.savedBytes = 0;
    envelope.savedRatio = 0;
  }

  return envelope;
}

export function decompressResult(envelope, { keyRegistry = null } = {}) {
  if (!envelope || typeof envelope !== 'object' || envelope.encoding !== 'lapc-json-v1') {
    throw new Error('Unsupported compression envelope');
  }
  if (!envelope.compressed) {
    return envelope.data;
  }
  let keys = envelope.keys;
  if ((!keys || typeof keys !== 'object') && typeof envelope.keysRef === 'string' && keyRegistry) {
    keys = keyRegistry[envelope.keysRef];
  }
  if (!keys || typeof keys !== 'object') {
    throw new Error('Invalid or missing key dictionary in envelope');
  }
  return decode(envelope.data, keys);
}

export function tokenSavings(original, candidate, counter = null) {
  const tc = counter || new TokenCounter();
  return tc.count(original) - tc.count(candidate);
}

export function estimateCompressibility(value) {
  const keyCounter = {};
  const scalarCounter = {};
  let homogeneousLists = 0;
  let totalLists = 0;

  const walk = (node) => {
    if (Array.isArray(node)) {
      totalLists += 1;
      if (node.length > 0 && node.every((item) => item && typeof item === 'object' && !Array.isArray(item))) {
        const keysets = node.map((item) => JSON.stringify(Object.keys(item).sort()));
        if (new Set(keysets).size === 1) {
          homogeneousLists += 1;
        }
      }
      for (const child of node) walk(child);
      return;
    }

    if (node && typeof node === 'object') {
      for (const [key, child] of Object.entries(node)) {
        const name = String(key);
        keyCounter[name] = (keyCounter[name] || 0) + 1;
        walk(child);
      }
      return;
    }

    if (typeof node === 'string' || typeof node === 'number' || typeof node === 'boolean' || node === null) {
      const marker = JSON.stringify(node);
      scalarCounter[marker] = (scalarCounter[marker] || 0) + 1;
    }
  };

  walk(value);

  const totalKeys = Object.values(keyCounter).reduce((acc, n) => acc + n, 0);
  const duplicateKeys = Math.max(0, totalKeys - Object.keys(keyCounter).length);
  const keyRepeatRatio = totalKeys > 0 ? duplicateKeys / totalKeys : 0;

  const totalScalars = Object.values(scalarCounter).reduce((acc, n) => acc + n, 0);
  const duplicateScalars = Math.max(0, totalScalars - Object.keys(scalarCounter).length);
  const scalarRepeatRatio = totalScalars > 0 ? duplicateScalars / totalScalars : 0;

  const homogeneousRatio = totalLists > 0 ? homogeneousLists / totalLists : 0;
  const score = 0.5 * keyRepeatRatio + 0.25 * scalarRepeatRatio + 0.25 * homogeneousRatio;
  if (score < 0) return 0;
  if (score > 1) return 1;
  return score;
}

